### ISSUE TITLE
Multipart mixed responses generate warnings

### ISSUE THREAD

**Issue by ndw** [urllib3/urllib3#800]
If you read a multipart/mixed response, the connectionpool issues a couple of warnings about defects in the message. I'm not sure what the expected, correct behavior is, but these warnings seem spurious.

Stick this perl script somewhere:

```
#!/usr/bin/perl

print "Server: Some Server Name\r\n";
print "Content-Type: multipart/mixed; boundary=36eeb8c4e26d842a\r\n";
print "Content-Length: 178\r\n";
print "\r\n\r\n";
print "--36eeb8c4e26d842a\r\n";
print "Content-Type: text/plain\r\n";
print "\r\n";
print "7\r\n";
print "--36eeb8c4e26d842a\r\n";
print "Content-Type: text/plain\r\n";
print "\r\n";
print "9\r\n";
print "--36eeb8c4e26d842a\r\n";
print "Content-Type: text/plain\r\n";
print "\r\n";
print "11\r\n";
print "--36eeb8c4e26d842a--\r\n";
```

Read it with requests (naturally, you'll have to change the URI to wherever you put the script):

```
import requests, logging

logging.basicConfig(level=logging.WARNING)
logging.getLogger("requests").setLevel(logging.DEBUG)

headers = {'accept': "multipart/mixed"}
r = requests.get("http://localhost:8124/cgi-bin/mpm.pl", headers=headers)

print(r)
```

The following errors are displayed:

```
DEBUG:requests.packages.urllib3.connectionpool:"GET http://localhost:8124/cgi-bin/mpm.pl HTTP/1.1" 200 178
WARNING:requests.packages.urllib3.connectionpool:Failed to parse headers (url=http://localhost:8888/http://localhost:8124/cgi-bin/mpm.pl): [StartBoundaryNotFoundDefect(), MultipartInvariantViolationDefect()], unparsed data: ''
Traceback (most recent call last):
  File "/home/ndw/.virtualenvs/pyapi/lib/python3.4/site-packages/requests-2.8.0-py3.4.egg/requests/packages/urllib3/connectionpool.py", line 390, in _make_request
    assert_header_parsing(httplib_response.msg)
  File "/home/ndw/.virtualenvs/pyapi/lib/python3.4/site-packages/requests-2.8.0-py3.4.egg/requests/packages/urllib3/util/response.py", line 58, in assert_header_parsing
    raise HeaderParsingError(defects=defects, unparsed_data=unparsed_data)
requests.packages.urllib3.exceptions.HeaderParsingError: [StartBoundaryNotFoundDefect(), MultipartInvariantViolationDefect()], unparsed data: ''
```

It took me quite a while to work out that they were spurious (because in real life, the server side that is generating the multipart/mixed is more complicated!)

See also kennethreitz/requests#3001
##

**Comment by shazow** [comment]
Just eyeballing it, but I _think_ the boundary line needs another \r\n at the end for it to be correct encoding.

```diff
- print "Content-Type: multipart/mixed; boundary=36eeb8c4e26d842a\r\n";
+ print "Content-Type: multipart/mixed; boundary=36eeb8c4e26d842a\r\n\r\n";
```

**Comment by shazow** [comment]
Actually I'm mistaken, nevermind.

**Comment by sigmavirus24** [comment]
Hey all, so I took a quick look at this by taking the text and shoving it into a text file (with a status-line) and serving it via netcat while requesting it with only plain `http.client`. I wasn't very familiar with [`assert_header_parsing`](https://github.com/shazow/urllib3/blob/c325f97b7e892cefe5676e129c35c9be5d8180cd/urllib3/util/response.py#L31) but the reason you're seeing this is because of http.client's parsing of the headers. Here's the code that brought me to this conclusion:

```py
>>> conn = http.client.HTTPConnection('localhost', 3234)
>>> conn.request('GET', '/')
>>> conn.getresponse()
<http.client.HTTPResponse object at 0x11000e6a0>
>>> r = _
>>> r.headers
<http.client.HTTPMessage object at 0x11003b0b8>
>>> r.headers.defects
[StartBoundaryNotFoundDefect(), MultipartInvariantViolationDefect()]
```

That said, this doesn't seem to be a problem for http.client or urllib3. This simply issues a log warning which might not be the right log level for this (since it isn't anything more than informational).

That said, @ndw, what would you consider to be an appropriate fix here?

**Comment by ndw** [comment]
Lowering the warning level so that they require an even "finer" logging level would be fine, I guess, if the conditions that are being reported really are harmless. They're noisy and confusing when the logging level is set to warning.

On the other hand, if I understood what the warnings meant, I might be equally content to amend my code so as not to raise them. However, AFAICT, there is a start boundary, so I don't know why it's "not found" and I couldn't find anything in the MIME RFC that seemed to correspond to a multipart invariancy.

**Comment by birm** [comment]
I think this traces all the way back to the email/rfc822 parser, and how it's trying to track your header.  in the python source, I don't think it gets to [line 355](https://github.com/python/cpython/blob/2d264235f6e066611b412f7c2e1603866e0f7f1b/Lib/email/feedparser.py#L365), thus it raises the error.


I think the issue is [here](https://github.com/python/cpython/blob/2d264235f6e066611b412f7c2e1603866e0f7f1b/Lib/email/feedparser.py#L343). Since it sees a blank line before it sees your boundary, it raises the error.

To fix, if you can call it that, try adding your before you print any blank lines on your perl script.

**Comment by Genestealer** [comment]
Is this issue still outstanding?

**Comment by Lukasa** [comment]
As far as I know, yes.

**Comment by McVillano** [comment]
Hello, I see some fixes, but in 0.77.3 I´m still having this issue, can you please say which files I must change and what to apply the fix?
Kind regards

**Comment by dsignr** [comment]
As of 2019, I can confirm this issue still exists, particularly while using `boto3` on AWS Lambda. It's very random to be honest.

**Comment by rizlas** [comment]
Same with azure [https://github.com/Azure/azure-storage-python/issues/167](https://github.com/Azure/azure-storage-python/issues/167)

**Comment by ecbftw** [comment]
This happens consistently when using pypsrp to transfer files.

**Comment by tyler-hitzeman** [comment]
This warning also happens with WebEx's NBR SOAP API

**Comment by wlongabaugh** [comment]
This also happens when reading from Google Healthcare API's DICOMWeb service.

**Comment by sethmlarson** [comment]
This issue has been solved in master and will be released in v1.26.0
